### ISSUE TITLE
media proxy redirects to non-existing files/(small|original)/missing.png

### ISSUE THREAD

**Issue by abochmann** [tootsuite/mastodon#14097]
When a request for a media file is handled by the media proxy, it sometimes returns a redirect to files/small/missing.png or files/original/missing.png, both of which don't exist in the Mastodon distribution.

### Expected behaviour

Media proxy should redirect to an actual file (I assume the remote fetch fails).

### Actual behaviour

Example from the journal:

`bundle[6406]: [6460a307-2edc-4e9b-815e-a80bc1cf58df] method=GET path=/media_proxy/2469306/original format=html controller=MediaProxyController action=show status=302 duration=18.91 view=0.00 db=2.46 location=https://mastodon.example.com/files/original/missing.png`

### Steps to reproduce the problem

I don't know how to actively provoke an occurrence, but looking at the log on a Mastodon instance with something like `journalctl -u mastodon-web | grep "MediaProxyController.*missing\.png"` should turn up appropriate examples.

### Specifications

Mastodon v3.1.4
(but I've first seen it like 8 months ago)

**Comment by abochmann** [comment]
This redirect possibly happens when the media proxy wants to fetch something from an instance that has its media blocked on the local instance.

**Comment by ThibG** [comment]
This is an awkward, but rather minor issue. This only occurs when the media file is not available and Mastodon won't fetch it upstream (because media from the upstream server is rejected?), which means you can only return a 404. I guess it would be cleaner to return a 404 directly rather than redirect to something that doesn't exist, though…

**Comment by colinstu12** [comment]
agreed @ShadowJonathan this is indeed becoming a bigger issue as time goes on. More servers. Longer periods of time. More media to build up. More chances of this occurring. 

I'm ok with the XY problem not being fixed as long as this one gets fixed. Ideally would fix it at this level right?

What I don't get is why these changes/updates to the media path aren't triggering out updates to instances? Edits to toots send out a message, why not the same for backend stuff like the image path?

**Comment by ShadowJonathan** [comment]
Because posts are effectively being implicitly edited, the media URL is generated at the time of fetch or push, so they could change depending on when that URL has gotten fetched.

**Comment by colinstu12** [comment]
> could change depending on when that URL has gotten fetched.

The media path has been found to be changing however. How/why is that change not sending an update message out? If a remote admin were to have their media URLs change, ideally all of those changes would be communicated out to instances that had already received the previous status/toot that is having its media URL updated right? Just like when a user edits a toot, they want that update to go out, would want the media path to follow that same kind of mechanism too. So everyone should have the latest info in their toot (including source media URL) and then it'd be able to properly refetch the media after having been previously purged.

**Comment by ShadowJonathan** [comment]
> If a remote admin were to have their media URLs change, ideally all of those changes would be communicated out to instances that had already received the previous status/toot that is having its media URL updated right

This would trigger an update for nearly every media post that that server would have made, which could range in the hundreds of thousands, multiplied by the followers that those users would have.

While kinda acceptable on small servers, for long-running big servers this would not be practical, and clog the push queue for days on end.

Having a fail-and-then-update algorithm on the receiving server could handle this problem, the changing of domains, and many more.

Often enough, that server would've either forgotten, or already cached the media previously, and so it would only care about the media URL changing at the time a user wishes to re-observe that post again, where it would then perform a fetch on the media, and *then* it would be a prudent time to re-fetch the post (with media attachments). Overall, it would be more efficient in a federated network, and only update data that is required.
